internal/controller/controllers/agent_controller.go

```diff
@@ -17,29 +17,193 @@ limitations under the License.
 type AgentReconciler struct {
 	client.Client
-	Log    logrus.FieldLogger
-	Scheme *runtime.Scheme
+	Log       logrus.FieldLogger
+	Scheme    *runtime.Scheme
+	Installer bminventory.InstallerInternals
 }
 
 // +kubebuilder:rbac:groups=adi.io.my.domain,resources=agents,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=adi.io.my.domain,resources=agents/status,verbs=get;update;patch
 
 func (r *AgentReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	// your logic here
+	ctx := context.Background()
+	agent := &adiiov1alpha1.Agent{}
+	var Requeue bool
+	var inventoryErr error
+
+	err := r.Get(ctx, req.NamespacedName, agent)
+	if err != nil {
+		r.Log.WithError(err).Errorf("Failed to get resource %s", req.NamespacedName)
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	if agent.Spec.ClusterDeploymentName == nil {
+		err = errors.New("Cluster Deployment Name not set")
+		r.updateFailure(ctx, agent, err)
+		return ctrl.Result{}, nil
+	}
+	kubeKey := types.NamespacedName{
+		Namespace: agent.Spec.ClusterDeploymentName.Namespace,
+		Name:      agent.Spec.ClusterDeploymentName.Name,
+	}
+	clusterDeployment := &hivev1.ClusterDeployment{}
+
+	// Retrieve clusterDeployment
+	if err = r.Get(ctx, kubeKey, clusterDeployment); err != nil {
+		errMsg := fmt.Sprintf("failed to get clusterDeployment with name %s in namespace %s",
+			agent.Spec.ClusterDeploymentName.Name, agent.Spec.ClusterDeploymentName.Namespace)
+		Requeue = false
+		clientError := true
+		if !k8serrors.IsNotFound(err) {
+			Requeue = true
+			clientError = false
+		}
+		clusterDeploymentRefErr := newKubeAPIError(errors.Wrapf(err, errMsg), clientError)
+		// Update that we failed to retrieve the clusterDeployment
+		r.updateFailure(ctx, agent, clusterDeploymentRefErr)
+		return ctrl.Result{Requeue: Requeue}, nil
+	}
+
+	// Retrieve cluster from the database
+	cluster, err := r.Installer.GetClusterByKubeKey(kubeKey)
+	if err != nil {
+		if gorm.IsRecordNotFoundError(err) {
+			Requeue = true
+			inventoryErr = common.NewApiError(http.StatusNotFound, err)
+		} else {
+			Requeue = false
+			inventoryErr = common.NewApiError(http.StatusInternalServerError, err)
+		}
+		// Update that we failed to retrieve the cluster from the database
+		r.updateFailure(ctx, agent, inventoryErr)
+		return ctrl.Result{Requeue: Requeue}, nil
+	}
+
+	var result ctrl.Result
+	// check for updates from user, compare spec and update if needed
+	result, err = r.updateIfNeeded(ctx, agent, cluster)
+	if err != nil {
+		r.updateFailure(ctx, agent, err)
+		return result, nil
+	}
+
+	conditionsv1.SetStatusCondition(&agent.Status.Conditions, conditionsv1.Condition{
+		Type:    adiiov1alpha1.AgentSyncedCondition,
+		Status:  corev1.ConditionTrue,
+		Reason:  adiiov1alpha1.AgentSyncedReason,
+		Message: adiiov1alpha1.AgentStateSynced,
+	})
+	if updateErr := r.Status().Update(ctx, agent); updateErr != nil {
+		r.Log.WithError(updateErr).Error("failed to update agent status")
```

Comment:
Should requeue in this state, backed was updated but the status is incorrect.

Reply:
Done.